### PR TITLE
DOC: Add section on executing notebooks

### PR DESCRIPTION
**Pull request by tritemio**
This section shows how to execute notebooks using nbconvert API. The executed notebook is saved in notebook format. There is no discussion on exporting notebooks to other format (topic already addressed in another section of the docs).

This is heavily based on this blog post: [Batch Execution of Jupyter Notebooks](http://tritemio.github.io/smbits/2016/01/02/execute-notebooks/).

Except for links/formatting fixes it should be ready for review. I can rebase to clean history before merge.

### CODE REVIEW

File: docs/source/execute_api.rst [hunk 1]
```diff
@@ -0,0 +1,145 @@
+Executing notebooks
+===================
+
+In this section we show how to execute a ``.ipynb`` notebook
+document saving the result in notebook format.
+To export the notebook to other formats see the
+:ref:`next section <nbconvert_library>`.
```
Comment (takluyver) [inline docs/source/execute_api.rst]: It's probably better to refer to this by name, because if we rearrange the docs or put another page in between, we won't think about this.

File: docs/source/execute_api.rst [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+Executing notebooks programmatically is useful, for example, as a test layer
+in python libraries that include example notebooks, or as a way to
+automate the data analysis in projects involving more than one notebook.
+
+The same functionality of executing notebooks is exposed through a
+:ref:`command line interface <commandline_usage>` or a python API interface.
+In this section we will (mostly) illustrate how to use the python API interface.
```
Comment (takluyver) [inline docs/source/execute_api.rst]: Since the title of the page is just 'executing notebooks', I think it should at least briefly mention the `--execute` command line flag.

File: docs/source/execute_api.rst [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+.. code-block:: python
+
+    ep = ExecutePreprocessor(timeout=3600, kernel_name='python3')
+
+We specified two (optional) arguments ``timeout`` and ``kernel_name``, which
+define respectively the execution timeout and the execution kernel.
+
+    The **kernel_name** keyword `requires <https://github.com/jupyter/nbconvert/pull/177>`__
+    nbconvert 4.2 (unreleased, use master branch from github).
```
Comment (takluyver) [inline docs/source/execute_api.rst]: It's better not to scatter things like this ("unreleased, use master") through the docs, because we're definitely going to forget to update them. Just say that it's new in 4.2.

Also, I wouldn't show the `kernel_name` parameter in the first example at all. If it's not there, nbconvert will take the kernel name as saved in the notebook, which should be a sensible default in most cases.

File: docs/source/execute_api.rst [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    ep.preprocess(nb, {'metadata': {'path': 'notebooks/'}})
+
+Hopefully, we will not get any errors during the notebook execution
+(see the last section for error handling). Note that ``path`` specifies
+in which folder to execute the notebook.
+Finally, to save the resulting notebook:
+
+.. code-block:: python
+
+    nbformat.write(nb, open('executed_notebook.ipynb', mode='wt'))
```
Comment (takluyver) [inline docs/source/execute_api.rst]: Let's use a context manager:

```python
with open('executed_notebook.ipynb', 'wt') as f:
    nbformat.write(nb, f)
```

I like to follow best practices in code examples for docs much more rigorously than I do in actual code files, because code examples will end up being copied and pasted everywhere.

File: docs/source/execute_api.rst [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+That's all. Your executed notebook will be saved in the current folder
+in the file *executed_notebook.ipynb*.
```
Comment (takluyver) [inline docs/source/execute_api.rst]: I think the filename should be `monospaced` rather than *italicised*

File: docs/source/execute_api.rst [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+Execution arguments
+-------------------
+
+The arguments passed to ``ExecutePreprocessor`` are configuration options
```
Comment (takluyver) [inline docs/source/execute_api.rst]: You can create cross links to the description of this class with syntax like this:

```
:class:`ExecutePreprocessor`
```

File: docs/source/execute_api.rst [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+class attributes. Moreover, each traitlet is automatically exposed
+as command-line options. For example, we can pass the timeout from the
+command-line like this::
+
+    jupyter nbconvert --ExecutePreprocessor.timeout=3600 --to notebook --execute mynotebook.ipynb
+
+Let's now discuss in more detail the two traitlets we used.
+
+The ``timeout`` traitlet defines the maximum time (in seconds) the notebook is
+allowed to run, if the execution takes longer an exception will be raised.
```
Comment (takluyver) [inline docs/source/execute_api.rst]: The timeout is per cell, not for the whole notebook

File: docs/source/execute_api.rst [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+the error and the original error name and message are also printed.
+After an error, we can still save the notebook as before:
+
+.. code-block:: python
+
+    nbformat.write(nb, open('executed_notebook.ipynb', mode='wt'))
+
+The saved notebook contains the output up until the failing cell,
+and includes a full stack-trace and error (which can help debugging).
+A pattern I use to execute notebooks while handling errors is the following:
```
Comment (takluyver) [inline docs/source/execute_api.rst]: I think we should reword this to avoid using "I" - it's fine for blog posts, but a bit confusing for docs.

File: docs/source/execute_api.rst [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+the saved notebook for further inspection.
+
+As a last scenario, sometimes notebooks contains independent computations
+in each code cell.
+In this case it can be useful to run the notebook until the end,
+in order to get a complete picture of all cells that are failing.
+Luckily enough, the ``allow_errors`` traitlet (default False) allows to do that.
+With ``allow_errors=True``,
+the notebook is executed until the end, and a ``CellExecutionError`` is raised
+if one or more cells threw an error. In this case, the output notebook
```
Comment (takluyver) [inline docs/source/execute_api.rst]: I don't think this is right. If allow_errors is True,  it runs through to the end, without throwing CellExecutionError at all. It's intended for notebooks that demonstrate an error for some reason, so an error is an expected part of their output.

I've just looked at the help string for `allow_errors`, and I think it has a confusing mistake - it says `True` when it should say `False`. I think it was initially designed as an `ignore_errors` flag and we forgot to update the help when we switched it.

Comment (tritemio) [inline docs/source/execute_api.rst]: Ops, yes, in fact due to the switched True/False I mis-interpreted it as raising an exception (but at the end) when `allow_errors=True`. I'm including a docstring fix in the updated PR.

**Comment by tritemio** [comment]
@takluyver, thanks for the comments. I fixed the the points you raised. 

Still, I left a note on `kernel_name` and regarding the command-line example (`--execute` flag), it is already there.

**Comment by tritemio** [comment]
Wait for a few fixes left.

**Comment by tritemio** [comment]
Done with fixes. 

- Locally, it builds with no warnings and the internal links work.
- Added example of command-line usage in the section intro
- Improved docstring for `timeout` and `allow_errors` traitlets
- Document `ExecutePreprocessors.process` and add it to API docs 
- Reworded docs index intro

Would be nice to add the `ExecutePreprocessor` traitlet docstrings to the API docs
but I don't know if it's possible using autodoc.

**Comment by tritemio** [comment]
I also I rebased to clean the history.

**Comment by willingc** [comment]
@tritemio Thanks for this PR. It's very helpful. I have some grammar and style edits to make the section more user friendly. :sunglasses:

File: docs/source/execute_api.rst [hunk 10]
```diff
@@ -0,0 +1,143 @@
+Executing notebooks
+===================
+
+.. module:: nbconvert.preprocessors
+
+In this section we show how to execute a ``.ipynb`` notebook
```
Comment (willingc) [inline docs/source/execute_api.rst]: Introduction paragraph:

Jupyter notebooks are often saved with cleared output cells. `nbconvert` provides a helpful way to execute the input cells of an `.ipynb` notebook file and save the results, both input and output cells, as an `.ipynb` file. This conversion can be very helpful when teaching using notebooks or exploring notebooks created by other researchers.

Comment (willingc) [inline docs/source/execute_api.rst]: Typo ...execute an ``.ipynb`` ...

File: docs/source/execute_api.rst [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+document saving the result in notebook format.
+To export the notebook to other formats see section
```
Comment (willingc) [inline docs/source/execute_api.rst]: ...notebook file to other formats, such as reStructured Text or Markdown, see ...

File: docs/source/execute_api.rst [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+in python libraries that include example notebooks, or as a way to
+automate the data analysis in projects involving more than one notebook.
+
+The same functionality of executing notebooks is exposed through a
+:doc:`command line interface <usage>` or a python API interface.
+As an example, a notebook can be executed from the command line with::
+
+    jupyter nbconvert --to notebook --execute mynotebook.ipynb
+
+In this section we will (mostly) illustrate how to use the python API interface.
```
Comment (willingc) [inline docs/source/execute_api.rst]: This section will illustrate ... the Python API interface.

File: docs/source/execute_api.rst [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+Next, we configure the notebook execution mode::
+
+    ep = ExecutePreprocessor(timeout=600, kernel_name='python3')
+
+We specified two (optional) arguments ``timeout`` and ``kernel_name``, which
+define respectively the execution timeout and the execution kernel.
+
+    The option to specify **kernel_name** is new in nbconvert 4.2.
+    When not specified or when using nbconvert <4.2,
+    the default python kernel is chosen.
```
Comment (willingc) [inline docs/source/execute_api.rst]: Python

File: docs/source/execute_api.rst [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+To actually run the notebook we call the method ``preprocess``::
+
+    ep.preprocess(nb, {'metadata': {'path': 'notebooks/'}})
+
+Hopefully, we will not get any errors during the notebook execution
+(see the last section for error handling). Note that ``path`` specifies
+in which folder to execute the notebook.
+Finally, to save the resulting notebook::
```
Comment (willingc) [inline docs/source/execute_api.rst]: Finally, ~~to~~ save the resulting notebook

File: docs/source/execute_api.rst [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+to be used for the execution. By default, the kernel name is obtained from the
+notebook metadata. The traitlet ``kernel_name`` allows to specify a user-defined
+kernel, overriding the value in the notebook metadata. A common use case
+is that of a python 2/3 library which includes documentation/testing
+notebooks. These notebooks will specify either a python2 or python3 kernel
+in their metadata
+(depending on the kernel used the last time the notebook was saved).
+In reality, these notebooks will work on both python 2/3 and, for testing,
+it is important to be able to execute them programmatically on both
+versions. Here the traitlet ``kernel_name`` comes to help:
```
Comment (willingc) [inline docs/source/execute_api.rst]: ~~comes to help~~ is helpful:

File: docs/source/execute_api.rst [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+we can just run each notebook twice, specifying first "python2" and then
+"python3" as kernel name.
+
+Error Handling
+--------------
+
+In the previous sections we saw how to save an executed notebook, assuming
+there are no execution error. But, what if there are errors?
```
Comment (willingc) [inline docs/source/execute_api.rst]: ... execution errors.

File: docs/source/execute_api.rst [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        msg = 'Error executing the notebook "%s".\n\n' % notebook_filename
+        msg += 'See notebook "%s" for the traceback.' % notebook_filename_out
+        print(msg)
+        raise
+    finally:
+        nbformat.write(nb, open(notebook_filename_out, mode='wt'))
+
+This will save the executed notebook regardless of execution errors.
+In case of errors, however, an additional message is printed and the
+``CellExecutionError`` is raised. The messages directs the user to
```
Comment (willingc) [inline docs/source/execute_api.rst]: The messages direct the

File: docs/source/execute_api.rst [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+the saved notebook for further inspection.
+
+As a last scenario, it is sometimes useful to execute notebooks which
+raise exceptions, for example to show an error conditions.
```
Comment (willingc) [inline docs/source/execute_api.rst]: ... an error condition.

File: docs/source/index.rst [hunk 19]
```diff
@@ -3,10 +3,12 @@ Converting Notebooks to Other Formats
 =====================================
 
 The ``nbconvert`` tool allows you to convert an ``.ipynb`` notebook
-document file into various static formats.
+document file into various static formats. It can also execute notebooks
+programmatically.
 
-Currently, ``nbconvert`` is provided as a command line tool, run as a script
-using Jupyter. It also powers the 'Download as' feature within the
+``nbconvert`` is both a python library and a command line tool (invoked as
```
Comment (willingc) [inline docs/source/index.rst]: Python

File: docs/source/execute_api.rst (hunk 10 above)
Comment (takluyver) [inline docs/source/execute_api.rst]: Interesting, I'd say "a dot-ipynb file", so using "an" looks odd to me

Comment (willingc) [inline docs/source/execute_api.rst]: @takluyver Good point. It's been done both ways throughout the docs. Since we do use ``.ipynb`` throughout the docs instead of ``ipynb`` file, I agree with you. Let's leave this as "a" ``.ipynb`` file.

File: docs/source/execute_api.rst [hunk 20]
```diff
@@ -0,0 +1,143 @@
+Executing notebooks
+===================
```
Comment (Carreau) [inline docs/source/execute_api.rst]: I think that this file conveys that manually creating an execute preprocessor is **the way** to execute notebook, which is not false, but not technically the canonical way. We have a special Exporter :`from nbconvert.exporters import NotebookExporter` which is made for notebook to notebook conversion (it does extra things). 

The canonical way would be

```python
from nbconvert.exporters import NotebookExporter, HTMLExporter
from traitlets.config import Config

c = Config({'ExecutePreprocessor':{'enable':True}})
E = NotebookExporter(config=c) # or E = HTMLExporter(config=c)

notebook, resources = E.from_filename('0 - notebook.ipynb')
# TODO save notebook
```

One of the big advantage of this approach is that it's identical wether you convert to HTML, rst, Latex, you "just" have to change the exporter you use. 

The other advantage is that it allows to imedialelty now how to do the same things for others and multiple preprocessors, in particular the following that all exporters inherit from : 

```python
List([
                                  'nbconvert.preprocessors.ClearOutputPreprocessor',
                                  'nbconvert.preprocessors.ExecutePreprocessor',
                                  'nbconvert.preprocessors.coalesce_streams',
                                  'nbconvert.preprocessors.SVG2PDFPreprocessor',
                                  'nbconvert.preprocessors.CSSHTMLHeaderPreprocessor',
                                  'nbconvert.preprocessors.LatexPreprocessor',
                                  'nbconvert.preprocessors.HighlightMagicsPreprocessor',
                                  'nbconvert.preprocessors.ExtractOutputPreprocessor',
                              ],
```

You "just" have to enable the preprocessors you use

The exporter should also take care of having the export process stateless, which make the conversion less prone to error. 

While the "how to use preprocessor" is nice, I think it should mention that unless you know what you are doing, using preprocessor directly is likely not what you want to do.

Comment (takluyver) [inline docs/source/execute_api.rst]: Hmm, I would disagree with you there. I think executing and converting are two separate steps, and if all you want to do is execute a notebook, you shouldn't need an exporter object involved. So I would consider using `ExecutePreprocessor` the more canonical approach.

Comment (Carreau) [inline docs/source/execute_api.rst]: Well the example give here is explicitly reading from disk, executing and writing back. So it is re-writing a basic exporter. Using the exporter also remove from many underlying details, like reading the notebook correctly (with nbformat) and setting the kernel to the right kerne, and then writing it back. 

Plus we know from experience that the order of preprocessors is tricky. So we likely do not want people to use preprocessor in the wrong way/wrong order and complain.

Comment (takluyver) [inline docs/source/execute_api.rst]: The notebook exporter only really exists to fit execution+saving into the framework of nbconvert. It's not really necessary, and I think it's unwanted complexity for this example. Even if it can result in slightly fewer lines of code, it obscures what's going on.

 ExecutePreprocessor already handles getting the kernel name from the notebook metadata, so Exporter is no help there. Reading a notebook file with nbformat is a) trivial, and b) definitely a public API, so it's good to show people how to use it.

I think preprocessor order has been problematic precisely because nbconvert separates enabling the preprocessors from their priority. I think using them explicitly as Python objects rather than enabling them through config makes it easier to get them organised correctly.

Comment (tritemio) [inline docs/source/execute_api.rst]: I agree with @takluyver here. In most of my use cases I need to execute notebooks and save the results as notebook (using `nbformat`, saving is not even handled by nbconvert).

Saving as notebook has 2 advantages for me. One is that the file can be distributed and is "executable". Compare this to a static representation in a sphinx documentation, that is harder to re-run (I should add a link to the original unexecuted notebook, but it's nice to have the output). Second, for debugging, it's easier to start from a notebook with all the output and re-execute it step by step, instead of starting from a  clean (no output) notebook. In this latter case, I use nbconvert for batch execution of a set of notebooks and when something fails I dig the problem by opening the failing notebook (containing output).

**Comment by tritemio** [comment]
@willingc I applied your fixes. Let me know if you want to further tweak the intro.

**Comment by tritemio** [comment]
@Carreau, I think your example on exporting notebooks should go on a different section, for example here:

- [Python API for working with nbconvert](http://nbconvert.readthedocs.org/en/latest/api/index.html)

However this section is currently a mess and should be heavily re-edited IMHO.

@willingc, @Carreau, @takluyver is there a way to make traitlets descriptions appear in sphinx API documentation?

**Comment by takluyver** [comment]
> is there a way to make traitlets descriptions appear in sphinx API documentation?

I don't think so, unfortunately. Maybe it could be done with a Sphinx extension, if somebody is interested in writing one.

**Comment by tritemio** [comment]
@takluyver, regarding the `with` statement, on a second though, I think in this case it may be ok not to use it. Take this example (reading):

```
nb = nbformat.read(open(notebook_filename), as_version=4)
```

No reference to the opened file is saved and the file is closed when `nbformat.read()` returns. The same goes for writing:

```
nbformat.write(nb, open('executed_notebook.ipynb', 'wt'))
```

And, in the final example on error handling, adding the `with` inside the `finally` becomes a little heavier on the eyes:

```
    try:
        out = ep.preprocess(nb, {'metadata': {'path': run_path}})
    except CellExecutionError:
        msg = 'Error executing the notebook "%s".\n\n' % notebook_filename
        msg += 'See notebook "%s" for the traceback.' % notebook_filename_out
        print(msg)
        raise
    finally:
        with open(notebook_filename_out, mode='wt') as f
            nbformat.write(nb, f)
```

What's your take on this?

**Comment by takluyver** [comment]
On Python 3, a file that isn't explicitly closed and gets cleaned up when all references to it are gone generates a ResourceWarning. Admittedly these aren't visible by default, and perhaps it's overly strict, but they do show up during tests, so I still think it's better to point people to `with open(...`.

**Comment by tritemio** [comment]
Ok, I didn't know that, I'll leave the PR as is then, using `with open(...`.

**Comment by tritemio** [comment]
@takluyver 

> ResourceWarning is ignored unless Python was built in debug mode.

https://docs.python.org/3.5/library/warnings.html#default-warning-filters

**Comment by takluyver** [comment]
Yep, like I said, they're not visible by default. You can enable them without rebuilding Python, though, and doing that when testing is good practice.

File: docs/source/execute_api.rst (hunk 20 above)
Comment (Carreau) [inline docs/source/execute_api.rst]: Fair enough.

**Comment by Carreau** [comment]
seeing that this is an improvement on our docs I would be +1 to merge and iterate.

**Comment by tritemio** [comment]
Regarding the verbosity of loading, executing and saving the notebook, I proposed in the past a top-level function https://github.com/jupyter/nbconvert/issues/125 to run notebooks in one step.

Nowadays I use a slightly improved version of that function which allows to optionally pass parameters to the notebook to be executed (~like a function). See [nbrun](https://github.com/tritemio/nbrun), rudimentary but incredibly useful in my workflow.

**Comment by willingc** [comment]
@Carreau +1 on merge and iterate. Will leave to you or @takluyver to merge. 

@tritemio Thank you! :cookie:

**Comment by tritemio** [comment]
@willingc, thanks, glad I earn a :cookie:!

**Comment by willingc** [comment]
@tritemio :+1: Also, :surfer: :palm_tree: :cake: is a virtual vacation ;)

**Comment by tritemio** [comment]
Just add one last commit, I'm done.

File: nbconvert/preprocessors/execute.py [hunk 21]
```diff
@@ -36,7 +36,13 @@ class ExecutePreprocessor(Preprocessor):
     """
 
     timeout = Integer(30, config=True,
-        help="The time to wait (in seconds) for output from executions."
+        help=dedent(
+            """
+            The time to wait (in seconds) for output from executions.
+            If a cell execution takes longer, a `CellExecutionError`
```
Comment (takluyver) [inline nbconvert/preprocessors/execute.py]: Looking at the code, it actually raises `TimeoutError` on Python 3 and `RuntimeError` on Python 2 (because TimeoutError is new in Python 3)

File: docs/source/index.rst (hunk 19 above)
Comment (takluyver) [inline docs/source/index.rst]: @willingc is right, it should be a capital P here.

File: nbconvert/preprocessors/execute.py (hunk 21 above)
Comment (tritemio) [inline nbconvert/preprocessors/execute.py]: @takluyver fixed in the last commit.

**Comment by takluyver** [comment]
Thanks for all your work on this!

**Comment by willingc** [comment]
Yes! Thanks @tritemio for the good work and @takluyver for the helpful reviews.

**Comment by Carreau** [comment]
:cake:

**Comment by tritemio** [comment]
:+1: @willingc @Carreau @takluyver
